### PR TITLE
feat: Allow additional routes for intra and private routing tables

### PR DESCRIPTION
**Pull request by bogdando**
## Description

Allow additional routes to be added for the private and intra routing tables, like it is already available for the default table.

An example use case could be peering with some VPC and making it routed with the VPC managed by the module. This example assumes that requesting and accepting the peering request should be handled by a user. As well as adding default routes on the other (peered VPC) side.

Ensure no conflicts for additional routing rules with private routes defined for NAT gateway managed with this module.

In order to support both NAT gateway routes on the private route tables, and the ones provided via the new parameters for custom routes, use an aws_route resource instead of the one for managing route tables. That is, per the documentation:

    NOTE on Route Tables and Routes: Terraform currently provides both
    a standalone Route resource and a Route Table resource with routes
    defined in-line. At this time you cannot use a Route Table with
    in-line routes in conjunction with any Route resources. Doing so
    will cause a conflict of rule settings and will overwrite rules.

To keep input values for custom default route table, private and intra routing rules consistent and interchangeble:
 * use the same aws_route approach for both private and intra cases;
 * alias (ipv6_)cidr_block and destination_(ipv6_)cidr_block so that both inputs can be used interchangebly for default route table and private/intra subnet route resources.

## Motivation and Context

The use case is configuring all required routes within the VPC module, instead of adding custom
calls for ``aws_route``. The latter is not always possible with terragrunt modules for example, which allows
only other modules there, and no custom resources.
(see https://github.com/gruntwork-io/terragrunt/issues/1187 and https://github.com/gruntwork-io/terragrunt/issues/1150)

Custom routes may be required for the private/intra VPC route tables to cover many cases, like
peering the managed VPC with another VPC (outside of this feature scope) and manage additional
routes on both sides. This change allows custom routes to be added for the managed VPC, for this
example. While adding the needed routes on the other side of the peering, and creating the peering would be
left for the user exercise.

Implements https://github.com/terraform-aws-modules/terraform-aws-vpc/issues/820

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects (with my [tEKS fork](https://github.com/bogdando/teks/blob/main/QUICKSTART.md#demo-dev-cluster-un-this-fork))

- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1041,6 +1041,77 @@ resource "aws_nat_gateway" "this" {
...
+  # consume it as end values. Iterate over a map with keys names matching
+  # router IDs (key unquieness is ensured by a hash of each rule)
+  # as for_each cannot take these from a list.
+  for_each = {
+    for elements in chunklist(flatten([
+      for pair in setproduct(
+        aws_route_table.private[*].id,
+        [var.private_routes_extra]
+      ) : setproduct([pair[0]], [for p in pair[1] : [jsonencode(p)]])
+    ]), 2) : "${elements[0]}:${sha256(elements[1])}" => jsondecode(elements[1])
```
Comment (antonbabenko) [inline main.tf:1056]: Sorry, but this code is not what we want to accept in this module...

**Review by antonbabenko — COMMENTED** [review]
We expect users to manage many other network-related resources outside of this module, even if such resources belong to VPC because there are a lot of variations and use cases.

We won't accept this PR because it is too complicated.

We may add support for `aws_route` resources in future major releases.

File: main.tf (hunk 1 above)
Comment (bogdando) [reply main.tf:1056]: Thank you for feedback. I couldn't find another approach to make setproduct and for_each working on different data types - strings for route tables IDs and lists of maps for roles. If you have ideas to share, I'd appreciate. I'm quite new to terraform, so please excuse my ignorance.

Comment (bogdando) [reply main.tf:1056]: Please also note linked terragrunt issues,. It seems there is no github URL to use aws_route as a dedicated module alongside vpc one

**Comment by bogdando** [comment]
I workarounded this in my teks fork by using [another module](https://github.com/bogdando/teks/commit/be2398422497d80c9c6d7340c2799967fb00cb98) to manage VPC peering and custom routes

**Pull request closed without merging**